src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java

```diff
@@ -1202,39 +1215,38 @@ public void run()
     @SuppressWarnings("resource")
     public SSTableReader cloneWithNewSummarySamplingLevel(ColumnFamilyStore parent, int samplingLevel) throws IOException
     {
-        synchronized (tidy.global)
-        {
-            assert openReason != OpenReason.EARLY;
+        assert openReason != OpenReason.EARLY;
 
-            int minIndexInterval = metadata().params.minIndexInterval;
-            int maxIndexInterval = metadata().params.maxIndexInterval;
-            double effectiveInterval = indexSummary.getEffectiveIndexInterval();
+        int minIndexInterval = metadata().params.minIndexInterval;
+        int maxIndexInterval = metadata().params.maxIndexInterval;
+        double effectiveInterval = indexSummary.getEffectiveIndexInterval();
```

Comment:
@bdeggleston @jasonstack I [mentioned](https://issues.apache.org/jira/browse/CASSANDRA-15861?focusedCommentId=17170484&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17170484) this briefly in the Jira, but with `indexSummary` not `final` or `volatile` in `SSTableReader`, do we have a visibility issue? The thread that loads/builds the summary needs to safely publish its changes for the thread that builds new summaries here.

Simply making `indexSummary` `volatile` might be acceptable. Most of its accesses are themselves part of the summary redistribution task itself or during compaction, so we probably don't care about that overhead. `SSTableReader#getIndexScanPosition()`, on the other hand, does look like it's used at least during some range reads.

We could go back to the old scope for the `synchronized` block, leaving `indexSummary` non-volatile. That would at least ensure visibility between the `cloneX()` methods. Even then, there are other things that call `SSTableReader#internalOpen()`, like `BigTableWriter.openEarly()` and `openFinal()`, that don't initialize `indexSummary` while holding the monitor lock. (Perhaps we could just synchronize to get the `IndexSummary` reference, rather than having to wait for `buildSummaryAtLevel()`, though.)

Is there actually a bug here independent of this patch? i.e. Do we actually need to make sure the block that includes setting `indexSummary` in `SSTableReader#internalOpen()` is performed in a `synchronized (tidy.global)` block? This bring up another level of puzzlement for me, which is why we ever synchronize on the non-final `global` field, rather than just `tidy`, which seems like it would actually make it safe to do what I've mentioned above.

Reply:
There are other places that write `indexSummary` without synchronization as well, namely `SSTableReader#load()` and `openForBatch()`.

Reply:
I would have to spend a while familiarising myself with this patch (and all present usages) to be sure, but this is probably fine - the plain write will piggy-back on whatever `volatile` write is used to publish the SSTable to other threads. I think in any cases we care about it _should_ piggy-back on `Tidy.setup`.  The only possible issue is if we modify the summary while the sstable is already published, and even in this case we're fine because `IndexSummary` is an immutable class (by which I mean its fields are all final), so the worst that can happen is that it is not immediately visible (technically the JMM doesn't guarantee it is ever visible to other threads, but in practice it cannot fail to eventually make it visible)

Reply:
There _is_ a volatile write to `setup` in `InstanceTidier#setup()` after `ifile`, `indexSummary`, et al. are set on the `SSTableReader`, but I'm not sure that helps us reason about the visibility of `indexSummary` for the thread that enters `cloneWithNewSummarySamplingLevel()`?

Unless there's some consequence I'm missing around making `global` final and assigning it at `InstanceTidier` (which would at least get the static analysis tools off our backs), I'd really like to make our synchronization policy around the non-final fields of the reader explicit (i.e. around `tidy.global`) rather than relying on more subtle reasoning around a volatile.

In terms of how that all affects this patch, the fact that `IndexSummary` is immutable should make it possible for us to synchronize only to get the `IndexSummary` reference, rather than having to wait for `buildSummaryAtLevel()`.

Reply:
So, for the result of `cloneWithNewSummaryAtLevel` to ever be visible to anybody, it must be published safely - in this case ordinarily through the DataTracker.View. So anything we do to provide visibility earlier than this is just providing is some nicer feels, which is fine, but as it happens we have several such volatile writes: the synchronised block when saving the new summary provides visibility, as does the write to `tidy.setup` (which is literally its purpose, iirc, though for the scope of `tidy`).

I am generally in favour of making more explicit links between a visibility requirement and where it is enforced, but there's definitely no need to introduce extra synchronised blocks - a simple write to a volatile property is sufficient, and in this case a comment indicating we're piggy-backing on any of these mentioned volatile writes would be sufficient - as would none of them, since we must publish the sstable safely anyway.

Reply:
(This class is definitely a mess)

Reply:
I've had a chance to talk it over w/ @belliottsmith and think about it, and tl;dr the changes here should be safe. The big thing I was missing is that the `SSTableReader` itself should already be published safely between threads, and if it is, the `indexSummary` reference should be fine (i.e. not null, which is the only danger, because it doesn't actually change after creation given we clone).

However, it might be helpful to briefly document this in `cloneWithNewSummarySamplingLevel()` so future readers might avoid some confusion. Better yet, we could finally just make these fields final in `SSTableReader`, avoiding all this confusion in the first place. I'm guessing the reason they aren't is that we added some layers of initialization over time, but deferring reader creation until we're ready seems like it could be achieved with a builder (or even without one).

CC @jasonstack

Reply:
+1, that would be a big improvement

Reply:
I made some refactor based on @maedhroz 's builder prototype: https://github.com/jasonstack/cassandra/pull/16. Now we have `BuilderForWriter`/`BuilderForRead`/`BuilderForBatch` and `IndexSummary` is `final`. We can make `dfile/ifile/bf` final as well, but didn't do it in this commit..

what do you guys think?

Reply:
@jasonstack I took a look and dropped a couple nits. Overall, it looks pretty straightforward, a largely organizational change that isolates building logic from the reader itself. I _do_ think that we should make `dfile` / `ifile` / `bf` final while we're here and have the opportunity. `first` and `last` seem a little more complicated, given the way the writer uses them during append, so no big deal if those stay mutable.

Reply:
updated the PR to make `dfile/ifile/bf` final.. `first/last` are used in `SSTable` and modified by writer.. I will leave them as they are..